File: tests/test_swarm_intelligence.py
import pytest
import asyncio
from zena_mode.arbitrage import SwarmArbitrator, ConsensusProtocol


@pytest.fixture
def arbitrator():
    return SwarmArbitrator(ports=[8001, 8002])


def test_protocol_selection(arbitrator):
    """Verify that different task types map to correct protocols."""
    assert arbitrator.select_protocol("factual") == ConsensusProtocol.CONSENSUS
    assert arbitrator.select_protocol("reasoning") == ConsensusProtocol.WEIGHTED_VOTE
    assert arbitrator.select_protocol("creative") == ConsensusProtocol.VOTING
    assert arbitrator.select_protocol("unknown") == ConsensusProtocol.HYBRID


def test_expert_specialization_prompts(arbitrator):
    """Verify that specialized system prompts exist for roles."""
    assert "security" in arbitrator.TASK_SYSTEM_PROMPTS
    assert "performance" in arbitrator.TASK_SYSTEM_PROMPTS
    assert "You are a security auditor" in arbitrator.TASK_SYSTEM_PROMPTS["security"]


@pytest.mark.asyncio
async def test_latency_sorting(arbitrator, mocker):
    """Verify that ports are sorted by latency."""
    arbitrator.latencies = {8001: 0.5, 8002: 0.1}
    arbitrator.ports = [8001, 8002]

    # Trigger the sort logic (normally in discover_swarm)
    arbitrator.ports.sort(key=lambda p: arbitrator.latencies.get(p, 999))

    assert arbitrator.ports == [8002, 8001]


def test_hallucination_penalty_logic(arbitrator):
    """Verify that the verification score affects selection logic."""
    # Mock a verification result
    # The logic should eventually use this to penalize the agent's reliability
    pass


if __name__ == "__main__":
    pytest.main([__file__])

File: rust_output/src/test_swarm_intelligence.rs
use anyhow::{Result, Context};
use crate::arbitrage::{SwarmArbitrator, ConsensusProtocol};
use std::collections::HashMap;
use tokio;

pub fn arbitrator() -> () {
    SwarmArbitrator(/* ports= */ vec![8001, 8002])
}

/// Verify that different task types map to correct protocols.
pub fn test_protocol_selection(arbitrator: String) -> () {
    // Verify that different task types map to correct protocols.
    assert!(arbitrator.select_protocol("factual".to_string()) == ConsensusProtocol.CONSENSUS);
    assert!(arbitrator.select_protocol("reasoning".to_string()) == ConsensusProtocol.WEIGHTED_VOTE);
    assert!(arbitrator.select_protocol("creative".to_string()) == ConsensusProtocol.VOTING);
    assert!(arbitrator.select_protocol("unknown".to_string()) == ConsensusProtocol.HYBRID);
}

/// Verify that specialized system prompts exist for roles.
pub fn test_expert_specialization_prompts(arbitrator: String) -> () {
    // Verify that specialized system prompts exist for roles.
    assert!(arbitrator.TASK_SYSTEM_PROMPTS.contains(&"security".to_string()));
    assert!(arbitrator.TASK_SYSTEM_PROMPTS.contains(&"performance".to_string()));
    assert!(arbitrator.TASK_SYSTEM_PROMPTS["security".to_string()].contains(&"You are a security auditor".to_string()));
}

/// Verify that ports are sorted by latency.
pub async fn test_latency_sorting(arbitrator: String, mocker: String) -> () {
    // Verify that ports are sorted by latency.
    arbitrator.latencies = HashMap::from([(8001, 0.5_f64), (8002, 0.1_f64)]);
    arbitrator.ports = vec![8001, 8002];
    arbitrator.ports.sort(/* key= */ |p| arbitrator.latencies.get(&p).cloned().unwrap_or(999));
    assert!(arbitrator.ports == vec![8002, 8001]);
}

/// Verify that the verification score affects selection logic.
pub fn test_hallucination_penalty_logic(arbitrator: String) -> () {
    // Verify that the verification score affects selection logic.
    // pass
}
